### ISSUE TITLE
Set default CMake build_type to 'Release'?

### ISSUE THREAD

**Issue by scheibelp** [spack/spack#5265]
See: #5261, #5258

Currently all CMake packages use 'RelWithDebInfo' as the default build_type. This can produce large artifacts with dubious benefits.

**Comment by svenevs** [comment]
@pramodk had some interesting suggestions with respect to debuggers, but it might require more machinery than is feasible.  I'm currently debugging what I believe is an armadillo bug, but their build system is an excellent example of how coordinating the source code and the debugger, regardless of downloading the same code, is infeasible.  The CMake script calls a configure, which then generates the _actual_ source code being compiled.  It's pretty sleek, but hard to reproduce without also re-installing.

More generally, just because you can download the source code, how can you ever know where the actual source code was?

I would love this kind of functionality, but it seems impossible to actually carry out.

**Comment by adamjstewart** [comment]
> I'm currently debugging what I believe is an armadillo bug

If you're having problems with Armadillo, you should check out #2061 and #2808. I wrote the Spack package for Armadillo and discovered a nasty installation bug. The developers were absolutely useless, so the bug may still be present.

**Comment by svenevs** [comment]
Short version: yes, it is the very same bug, reproducible outside of spack.  It's their macro expansions.  I sent them an e-mail, but will probably receive the same response (didn't even search spack because it was clearly a bug with armadillo).  I may investigate more...I may not.  I'll post back on those if I get more info.

**Comment by pramodskumbhar** [comment]
>More generally, just because you can download the source code, how can you ever know where the actual source code was?
>I would love this kind of functionality, but it seems impossible to actually carry out.

I briefly mentioned this on Slack. Here is the example: I build `mod2c` package (local) with debug symbols using Spack and spack has removed source. If we try to debug:

```
○ → gdb /some-path/SPACK_INSTALL/viz/simulation/install/linux-rhel6-x86_64/gcc-4.4.7/mod2c-develop-oqmbuquoqfdbfh2pvrknwo6gtqbg5qcm/bin/mod2c_core
GNU gdb (GDB) Red Hat Enterprise Linux (7.2-92.el6)
…
Reading symbols from /some-path/SPACK_INSTALL/viz/simulation/install/linux-rhel6-x86_64/gcc-4.4.7/mod2c-develop-oqmbuquoqfdbfh2pvrknwo6gtqbg5qcm/bin/mod2c_core...done.

(gdb) show directories
Source directories searched: $cdir:$cwd

(gdb) b main
Breakpoint 1 at 0x412240: file /some-extra-path/spack/var/spack/stage/mod2c-develop-oqmbuquoqfdbfh2pvrknwo6gtqbg5qcm/mod2c/src/mod2c_core/modl.c, line 108.

(gdb) r
Starting program: /some-path/SPACK_INSTALL/viz/simulation/install/linux-rhel6-x86_64/gcc-4.4.7/mod2c-develop-oqmbuquoqfdbfh2pvrknwo6gtqbg5qcm/bin/mod2c_core

Breakpoint 1, main (argc=1, argv=0x7fffffffd758)
    at /some-extra-path/spack/var/spack/stage/mod2c-develop-oqmbuquoqfdbfh2pvrknwo6gtqbg5qcm/mod2c/src/mod2c_core/modl.c:108
108	/some-extra-path/spack/var/spack/stage/mod2c-develop-oqmbuquoqfdbfh2pvrknwo6gtqbg5qcm/mod2c/src/mod2c_core/modl.c: No such file or directory.
	in /some-extra-path/spack/var/spack/stage/mod2c-develop-oqmbuquoqfdbfh2pvrknwo6gtqbg5qcm/mod2c/src/mod2c_core/modl.c
Missing separate debuginfos, use: debuginfo-install glibc-2.12-1.209.el6_9.2.x86_64
```

If we know the version etc, we can just do : 

```
(gdb) set substitute-path /some-extra-path/spack/var/spack/stage/mod2c-develop-oqmbuquoqfdbfh2pvrknwo6gtqbg5qcm/mod2c /some-path/sources/mod2c
(gdb) n
115		if (argc > 1 && strcmp(argv[1], "--version") == 0) {
(gdb) n
129		init();			/* keywords into symbol table, initialize
(gdb) n
128		modprefix = prefix_;
(gdb)
```
and  get what we needed. 

Hence in development and deployment workflow I still find `RelWithDebInfo` very useful.

**Comment by alalazo** [comment]
##### TLDR

A summary of my arguments in support of using `Release` as default:

1. debugging a `RelWithDebInfo` software may be a pain (and not so useful due to compiler optimizations)
2. we'll be wasting some disk space due to debug info (which sometimes is relevant, like in the `llvm` case)

More generally I think that for any build system we should default to installations that are not inserting debug symbols and are compiled in "release" mode.

Regardless of the default we'll choose, I think it would be nice if for anything that asks for debug symbols we could copy the stage directory in `<prefix>/.src` or similar and patch in some way the dwarf section to point to that directory. This would probably build on the work already done by @junghans in #4102.

##### Longer explanation

For 1. I must say that maybe it's just my personal experience, but the first thing I usually do when I see a production bug on a C++ app is trying to reproduce it on a debug version (`-O0 -g` or lately `-Og`). The unfortunate cases where the bug occurs only in release mode are usually not the most pleasing debugging experiences one may have.

For 2. the mileage may vary depending on the packages. I tried to see the impact of debug symbols on 2 packages to have an idea:
```console
$ spack find -p openmpi
==> 2 installed packages.
-- linux-ubuntu14.04-x86_64 / gcc@4.8 ---------------------------
    openmpi@2.1.1  /home/mculpo/PycharmProjects/spack/opt/spack/linux-ubuntu14.04-x86_64/gcc-4.8/openmpi-2.1.1-wbyfkrrj2m35owiiavkktcfmu6bivsnu
    openmpi@2.1.1  /home/mculpo/PycharmProjects/spack/opt/spack/linux-ubuntu14.04-x86_64/gcc-4.8/openmpi-2.1.1-5csqo5vovfbiroy3dl6lyj2yrrm7la7g

$ du -sh /home/mculpo/PycharmProjects/spack/opt/spack/linux-ubuntu14.04-x86_64/gcc-4.8/openmpi-2.1.1-wbyfkrrj2m35owiiavkktcfmu6bivsnu /home/mculpo/PycharmProjects/spack/opt/spack/linux-ubuntu14.04-x86_64/gcc-4.8/openmpi-2.1.1-5csqo5vovfbiroy3dl6lyj2yrrm7la7g
26M	/home/mculpo/PycharmProjects/spack/opt/spack/linux-ubuntu14.04-x86_64/gcc-4.8/openmpi-2.1.1-wbyfkrrj2m35owiiavkktcfmu6bivsnu
129M	/home/mculpo/PycharmProjects/spack/opt/spack/linux-ubuntu14.04-x86_64/gcc-4.8/openmpi-2.1.1-5csqo5vovfbiroy3dl6lyj2yrrm7la7g

$ spack find -p hdf5
==> 2 installed packages.
-- linux-ubuntu14.04-x86_64 / gcc@4.8 ---------------------------
    hdf5@1.10.1  /home/mculpo/PycharmProjects/spack/opt/spack/linux-ubuntu14.04-x86_64/gcc-4.8/hdf5-1.10.1-rh62esj7salbazymxh25vllami3o3ehd
    hdf5@1.10.1  /home/mculpo/PycharmProjects/spack/opt/spack/linux-ubuntu14.04-x86_64/gcc-4.8/hdf5-1.10.1-fa5xpyl3nfpebwytheq3ck4llhfzqcom

$ du -sh /home/mculpo/PycharmProjects/spack/opt/spack/linux-ubuntu14.04-x86_64/gcc-4.8/hdf5-1.10.1-rh62esj7salbazymxh25vllami3o3ehd /home/mculpo/PycharmProjects/spack/opt/spack/linux-ubuntu14.04-x86_64/gcc-4.8/hdf5-1.10.1-fa5xpyl3nfpebwytheq3ck4llhfzqcom
54M	/home/mculpo/PycharmProjects/spack/opt/spack/linux-ubuntu14.04-x86_64/gcc-4.8/hdf5-1.10.1-rh62esj7salbazymxh25vllami3o3ehd
95M	/home/mculpo/PycharmProjects/spack/opt/spack/linux-ubuntu14.04-x86_64/gcc-4.8/hdf5-1.10.1-fa5xpyl3nfpebwytheq3ck4llhfzqcom
```

In the end I don't see a lot of benefits in having debug symbols by default in release builds - they are wasting space and if I had to debug something I would probably start by building a debug version of the application (not a release + debug symbols).

Finally:

> If we have source and know the version etc, we can just do :

"etc" means knowing exactly what Spack did (patches applied, commit hash retrieved, or similar). I wouldn't recommend doing this because if you don't put the same sources you used in compilation you'll end up mapping the execution to the wrong line of the file. And I believe that could cause more headaches than it cures.

**Comment by pramodskumbhar** [comment]
Agree about the storage space (and debug symbols might be unnecessary for the many packages).

Other things depend on the context (e.g. jobs running at scale with long queues, in case of failure having debug info is helpful even if it's not precise due to optimisations).

Does `build_type` go in `packages.yaml` as a preference? Depending upon users of packages (on a particular hpc cluster) one can customise which packages should be  with debug vs no-debug (or using compiler flags). So not an issue if we make `Release` a `default`.

**Comment by svenevs** [comment]
As per previous discussions I must necessarily default to agreeing with @alalazo :wink:

In particular, anything that got patched makes knowing the path potentially more misleading than helpful.  To always get the right location, you would always have to re-stage (I believe).

That said,

> Does `build_type` go in `packages.yaml` as a preference?

That would be excellent, but I don't think the existing setup allows this.  I was only able to get this to work by setting a specific packages' `variants`, only after knowing it was a `CMakePackage`.

**Comment by pramodskumbhar** [comment]
> I was only able to get this to work by setting a specific packages' variants

forgot about `variants` aspect. this is fine!

**Comment by jrood-nrel** [comment]
So my application Nalu, which is built on top of Trilinos, which are both `CMakePackage`, experience huge binary size with `RelWithDebInfo` as well because Trilinos is huge and linked statically. I think I will have to vote for changing the default to `Release`.

**Comment by alalazo** [comment]
Closing as stale, added a link in #15848 to remind us to explain how one could obtain the same effect by writing an appropriate `packages.yaml`.
